internal/neofs/tree.go

```diff
@@ -0,0 +1,279 @@
+	results := make(map[string]string, 3)
+
+	results[systemNameKV] = settingsFileName
+	results[versioningEnabledKV] = strconv.FormatBool(settings.VersioningEnabled)
+	results[lockConfigurationKV] = encodeLockConfiguration(settings.LockConfiguration)
+
+	return results
+}
+
+func (c *TreeClient) getSystemNode(ctx context.Context, cnrID *cid.ID, treeID string, path, meta []string) (*Node, error) {
+	return c.getNode(ctx, cnrID, treeID, systemNameKV, path, meta)
+}
+
+func (c *TreeClient) getRegularNode(ctx context.Context, cnrID *cid.ID, treeID string, path, meta []string) (*Node, error) {
+	return c.getNode(ctx, cnrID, treeID, fileNameKV, path, meta)
+}
+
+func (c *TreeClient) getNode(ctx context.Context, cnrID *cid.ID, treeID, pathAttr string, path, meta []string) (*Node, error) {
+	nodes, err := c.getNodes(ctx, cnrID, treeID, pathAttr, path, meta)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get nodes: %w", err)
+	}
+	if len(nodes) == 0 {
+		return nil, layer.ErrNotFound
+	}
+	if len(nodes) != 1 {
+		return nil, fmt.Errorf("found more than one node")
+	}
+
+	return newNode(nodes[0]), nil
+}
+
+func (c *TreeClient) getNodes(ctx context.Context, cnrID *cid.ID, treeID, pathAttr string, path, meta []string) ([]*tree.GetNodeByPathResponse_Info, error) {
+	request := &tree.GetNodeByPathRequest{
+		Body: &tree.GetNodeByPathRequest_Body{
+			ContainerId:   cnrID.ToV2().GetValue(),
+			TreeId:        treeID,
+			Path:          path,
+			Attributes:    meta,
+			PathAttribute: pathAttr,
+		},
+	}
+
+	resp, err := c.service.GetNodeByPath(ctx, request)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get node path: %w", err)
+	}
+
+	return resp.GetBody().GetNodes(), nil
+}
+
+func (c *TreeClient) addNode(ctx context.Context, cnrID *cid.ID, treeID string, parent uint64, meta map[string]string) (uint64, error) {
+	request := &tree.AddRequest{
+		Body: &tree.AddRequest_Body{
+			ContainerId: cnrID.ToV2().GetValue(),
+			TreeId:      treeID,
+			ParentId:    parent,
+			Meta:        metaToKV(meta),
+		},
+	}
+
+	resp, err := c.service.Add(ctx, request)
+	if err != nil {
+		return 0, err
+	}
+
+	return resp.GetBody().GetNodeId(), nil
+}
+
+func (c *TreeClient) moveNode(ctx context.Context, cnrID *cid.ID, treeID string, nodeID, parentID uint64, meta map[string]string) error {
+	request := &tree.MoveRequest{
+		Body: &tree.MoveRequest_Body{
+			ContainerId: cnrID.ToV2().GetValue(),
+			TreeId:      treeID,
+			NodeId:      nodeID,
+			ParentId:    parentID,
+			Meta:        metaToKV(meta),
+		},
+	}
+
+	_, err := c.service.Move(ctx, request)
+	return err
+
+}
+
+func metaToKV(meta map[string]string) []*tree.KeyValue {
+	result := make([]*tree.KeyValue, 0, len(meta))
+
+	for key, value := range meta {
+		result = append(result, &tree.KeyValue{Key: key, Value: []byte(value)})
+	}
+
+	return result
+}
+
+func parseLockConfiguration(value string) (*data.ObjectLockConfiguration, error) {
```

Comment:
It seems that we have some strict encoding rules. Let's add some tests to stress that it is important.